docs/src/main/asciidoc/extension-writing-dev-service.adoc

```diff
@@ -0,0 +1,123 @@
+////
+This document is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+////
+[id="extension-writing-dev-service"]
+= Writing a dev service
+include::_attributes.adoc[]
+:categories: writing-extensions
+:diataxis-type: howto
+:topics: extensions
+////
+////
+
+
+== Prerequisites
+
+- You should already have an xref:building-my-first-extension.adoc[extension structure] in place
+- You should have a containerised version of your external service (not all dev services rely on containers, but most do)
+
+== Creating a dev service
+
+If your extension provides APIs for connecting to an external service, it's a good idea to provide a xref:dev-services.adoc[dev service] implementation.
+
+To create a dev service, add a new build step into the extension processor class that returns a `DevServicesResultBuildItem`.
+Here, the link:https://hub.docker.com/_/hello-world`hello-world` image is used, but you should set up the right image for your service.
+
+[source%nowrap,java]
+----
+    @BuildStep(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class) {
+    public DevServicesResultBuildItem createContainer() {
+        DockerImageName dockerImageName = DockerImageName.parse("hello-world");
+        GenericContainer container = new GenericContainer<>(dockerImageName)
+                                         .withExposedPorts(SERVICE_PORT, OTHER_SERVICE_PORT)
+                                         .waitingFor(Wait.forLogMessage(".*" + "Started" + ".*", 1))
+                                         .withReuse(true);
+
+        container.start();
+
+        String newUrl = "http://" + container.getHost() + ":" + container.getMappedPort(SERVICE_PORT);
+        Map<String, String> configOverrides = Map.of("some-service.base-url", newUrl);
+
+        return new DevServicesResultBuildItem.RunningDevService(FEATURE, container.getContainerId(),
+            container::close, configOverrides)
+            .toBuildItem();
+    }
+----
+
+With this code, you should be able to see your container starting if you add your extension to a test application and run `quarkus dev`.
+However, the application will not be able to connect to it, because no ports are exposed. To expose ports, add `withExposedPorts` to the container construction.
+For example,
+
+[source%nowrap,java]
+----
+GenericContainer container = new GenericContainer<>(dockerImageName)
+                                 .withExposedPorts(SERVICE_PORT, OTHER_SERVICE_PORT);
+----
+
+Testcontainers will map these ports to random ports on the host. This avoids port conflicts, but presents a new problem – how do applications connect to the service in the container?
+
+To allow applications to connect, the extension should override the default configuration for the service with the mapped ports.
+This must be done after starting the container.
+For example,
```

Comment:
Unfortunately, this is not as simple when dealing with the shared network config. But this thing is complex and a bit in motion as I'm not very satisfied with what we have so probably better to keep it simple for now.

Reply:
I have to say, I looked at a lot of our core dev services as examples and the complexity was a bit overwhelming. I had a hard time deciding where the right balance was between a naive 4-line dev service like the ones we live-code in workshops, and the 60-LoC monsters we have for a lot of our services. We don't want to give people instructions that don't work, but we also don't want to make people decide that writing a dev service is the thing they want to attempt. :)

Reply:
Yeah, I think we need to find a way to hide the complexity better.

I'm not entirely sure it's going to be easy but we really need to.